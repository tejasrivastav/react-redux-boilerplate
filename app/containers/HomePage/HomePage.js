/*
 * HomePage
 *
 * This is the first thing users see of our App, at the '/' route
 */

import React from 'react';
import PropTypes from 'prop-types';
import ReposList from 'components/ReposList';
import './style.scss';

import {
  Tab, Tabs, TabList, TabPanel
} from 'react-tabs';
import 'react-tabs/style/react-tabs.css';

export default class HomePage extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function
  /**
   * when initial state username is not null, submit the form to load repos
   */
  componentDidMount() {
    const { loadPosts } = this.props;
    loadPosts();
  }

  render() {
    const {
      loading, error, repos, changeTab, activeTab, categories
    } = this.props;
    const reposListProps = {
      loading,
      error,
      repos
    };

    const tabs = categories.map((category) => ({
      label: category,
      content: (<div></div>)
    }));
    tabs[activeTab].content = (
      <ReposList {...reposListProps} />
    );

    /* eslint-disable react/no-array-index-key */
    return (
      <article>
        <div className="home-page">
          <section className="centered">
            <Tabs onSelect={changeTab}>
              <TabList>
                {
                  tabs.map((tab, idx) => <Tab key={idx}>{tab.label}</Tab>)
                }
              </TabList>
              {
                tabs.map((tab, idx) => <TabPanel key={idx}>{tab.content}</TabPanel>)
              }
            </Tabs>
          </section>
        </div>
      </article>
    );
  }
}

HomePage.propTypes = {
  loading: PropTypes.bool,
  error: PropTypes.oneOfType([PropTypes.object, PropTypes.bool]),
  repos: PropTypes.oneOfType([PropTypes.array, PropTypes.bool]),
  loadPosts: PropTypes.func,
  changeTab: PropTypes.func,
  activeTab: PropTypes.number,
  categories: PropTypes.array
};
