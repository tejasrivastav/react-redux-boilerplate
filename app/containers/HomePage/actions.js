/*
 * Home Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  CHANGE_USERNAME, CHANGE_TAB, DELETE_POST, RELOAD_POSTS, UPDATE_QUERY, SEARCH_PERFORMED
} from './constants';

/**
 * Changes the input field of the form
 *
 * @param  {name} name The new text of the input field
 *
 * @return {object}    An action object with a type of CHANGE_USERNAME
 */
export function changeUsername(name) {
  return {
    type: CHANGE_USERNAME,
    name
  };
}

export function changeTab(index) {
  return {
    type: CHANGE_TAB,
    index
  };
}

export function deletePost(index) {
  return {
    type: DELETE_POST,
    index
  };
}

export function reloadPosts() {
  return {
    type: RELOAD_POSTS
  };
}

export function updateQuery(text) {
  return {
    type: UPDATE_QUERY,
    text
  };
}

export function searchPerformed(results) {
  return {
    type: SEARCH_PERFORMED,
    results
  }
}
