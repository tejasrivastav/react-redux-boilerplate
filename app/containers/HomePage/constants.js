/*
 * HomeConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

export const CHANGE_USERNAME = 'boilerplate/Home/CHANGE_USERNAME';
export const CHANGE_TAB = 'boilerplate/Home/CHANGE_TAB';
export const DELETE_POST = 'boilerplate/Home/DELETE_POST';
export const RELOAD_POSTS = 'boilerplate/Home/RELOAD_POSTS';
