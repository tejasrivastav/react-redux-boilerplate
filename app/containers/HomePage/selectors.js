import { createSelector } from 'reselect';

const selectHome = (state) => state.home;

const makeSelectUsername = () => createSelector(
  selectHome,
  (homeState) => homeState.username
);

const makeSelectTab = () => createSelector(
  selectHome,
  (homeState) => homeState.activeIndex
);

export {
  selectHome,
  makeSelectUsername,
  makeSelectTab
};
