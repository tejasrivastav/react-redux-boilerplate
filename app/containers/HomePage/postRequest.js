import request from 'utils/request';

import { categories } from 'containers/App/constants';
/*
* The transformer function to add catgories to the reponse
* register the catgeories you want to add
* exectute them over the row
*/

function transform(repos) {
  repos.forEach((row) => {
    categories.forEach((category) => {
      if (category.exec(row)) {
        row.category = row.category || [];
        row.category.push(category.name);
      }
    });
  });
  return repos;
}

export default function (endpoint) {
  return request(endpoint).then(transform).catch((err) => {
    console.error(err);
  });
}
