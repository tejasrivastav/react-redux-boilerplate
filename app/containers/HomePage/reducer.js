import { CHANGE_USERNAME, CHANGE_TAB } from './constants';

// The initial state of the App
const initialState = {
  username: '',
  activeIndex: 0
};

function homeReducer(state = initialState, action) {
  switch (action.type) {
    case CHANGE_USERNAME:
      // Delete prefixed '@' from the github username
      return { ...state, username: action.name.replace(/@/gi, '') };
    case CHANGE_TAB:
      return { ...state, activeIndex: action.index };
    default:
      return state;
  }
}

export default homeReducer;
