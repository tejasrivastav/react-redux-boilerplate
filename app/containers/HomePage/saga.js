/**
 * Gets the repositories of the user from Github
 */

import {
  call, put, takeLatest, all, select, debounce
} from 'redux-saga/effects';

import { LOAD_POSTS } from 'containers/App/constants';
import { postsLoaded, postLoadingError } from 'containers/App/actions';
import { selectCategoryPosts } from 'containers/App/selectors';
import request from 'containers/HomePage/services/postRequest';
import { search } from 'containers/HomePage/services/search';
import { UPDATE_QUERY } from './constants';
import { selectQuery } from './selectors';
import { searchPerformed } from './actions';

/**
 * Posts request/response handler
 */
export function* getPosts() {
  const endpoint = '/posts';

  try {
    const repos = yield call(request, endpoint);
    yield put(postsLoaded(repos));
  } catch (err) {
    yield put(postLoadingError(err));
  }
}

export function* performSearch() {
  try {
    const query = yield select(selectQuery);
    const posts = yield select(selectCategoryPosts);
    if (query.trim().length > 0) {
      const results = yield call(search, posts, query);
      yield put(searchPerformed(results));
    } else {
      yield put(searchPerformed([]));
    }
  } catch (err) {
    console.log(err);
    // TODO: handle error
  }
}
/**
 * Root saga manages watcher lifecycle
 */
export default function* getData() {
  // Watches for LOAD_POSTS actions and calls getPosts when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield all([
    takeLatest(LOAD_POSTS, getPosts),
    debounce(200, UPDATE_QUERY, performSearch)
  ]);
}
