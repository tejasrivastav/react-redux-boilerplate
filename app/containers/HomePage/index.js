import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';
import injectReducer from 'utils/injectReducer';
import injectSaga from 'utils/injectSaga';
import {
  makeSelectPosts,
  makeSelectLoading,
  makeSelectError,
  makeSelectCategories
} from 'containers/App/selectors';
import { loadPosts } from 'containers/App/actions';

import { changeUsername, changeTab } from './actions';
import { makeSelectUsername, makeSelectTab } from './selectors';
import reducer from './reducer';
import saga from './saga';
import HomePage from './HomePage';

const mapDispatchToProps = (dispatch) => ({
  onChangeUsername: (evt) => dispatch(changeUsername(evt.target.value)),
  loadPosts: (evt) => {
    if (evt !== undefined && evt.preventDefault) evt.preventDefault();
    dispatch(loadPosts());
  },
  changeTab: (index) => {
    dispatch(changeTab(index));
  }
});

const mapStateToProps = createStructuredSelector({
  repos: makeSelectPosts(),
  username: makeSelectUsername(),
  loading: makeSelectLoading(),
  error: makeSelectError(),
  activeTab: makeSelectTab(),
  categories: makeSelectCategories()
});

const withConnect = connect(mapStateToProps, mapDispatchToProps);

const withReducer = injectReducer({ key: 'home', reducer });
const withSaga = injectSaga({ key: 'home', saga });

export default compose(withReducer, withSaga, withConnect)(HomePage);
export { mapDispatchToProps };
