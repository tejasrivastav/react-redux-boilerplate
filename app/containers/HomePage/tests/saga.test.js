/**
 * Tests for HomePage sagas
 */

import {
  put, takeLatest, all, debounce, select
} from 'redux-saga/effects';
import { expectSaga } from 'redux-saga-test-plan';

import { LOAD_POSTS } from 'containers/App/constants';
import { selectCategoryPosts } from 'containers/App/selectors';
import { postsLoaded, postLoadingError } from 'containers/App/actions';
import { UPDATE_QUERY } from 'containers/HomePage/constants';
import { selectQuery } from 'containers/HomePage/selectors';

import { searchPerformed, updateQuery } from 'containers/HomePage/actions';
import getData, { getPosts, performSearch } from '../saga';


/* eslint-disable redux-saga/yield-effects */

describe('getDataSaga Saga', () => {
  const getDataSaga = getData();

  it('should start task to watch for LOAD_POSTS and UPDATE_QUERY actions', () => {
    const takeLatestDescriptor = getDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(all([
      takeLatest(LOAD_POSTS, getPosts),
      debounce(200, UPDATE_QUERY, performSearch)
    ]));
  });
});

describe('getPosts Saga', () => {
  let getPostsGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getPostsGenerator = getPosts();

    const selectDescriptor = getPostsGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();
  });

  it('should dispatch the postsLoaded action if it requests the data successfully', () => {
    const response = [{
      name: 'First repo',
    }, {
      name: 'Second repo',
    }];
    const putDescriptor = getPostsGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(postsLoaded(response)));
  });

  it('should call the postLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getPostsGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(postLoadingError(response)));
  });
});

describe('Perform Search', () => {
  it('should return the indices of matched results', () => {
    const query = 'First';
    const posts = [{
      id: 1,
      title: 'First post'
    }, {
      id: 2,
      title: 'Second post'
    }];
    const response = [{ id: 1, indexes: [[0, 5]] }];

    return expectSaga(performSearch)
      .provide([
        [select(selectQuery), query],
        [select(selectCategoryPosts), posts]
      ])
      .put(searchPerformed(response))
      .dispatch(updateQuery(query))
      .run();
  });

  it('should return the empty array for empty query', () => {
    const query = '';
    const posts = [{
      id: 1,
      title: 'First post'
    }, {
      id: 2,
      title: 'Second post'
    }];
    const response = [];

    return expectSaga(performSearch)
      .provide([
        [select(selectQuery), query],
        [select(selectCategoryPosts), posts]
      ])
      .put(searchPerformed(response))
      .dispatch(updateQuery(query))
      .run();
  });
});
