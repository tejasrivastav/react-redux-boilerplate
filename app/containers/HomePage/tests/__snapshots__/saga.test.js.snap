// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPosts Saga should call the postLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "/posts",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getPosts Saga should call the postLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "posts": "flexdinesh",
      "type": "boilerplate/App/LOAD_POSTS_SUCCESS",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`getPosts Saga should dispatch the postsLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "/posts",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getPosts Saga should dispatch the postsLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "posts": "flexdinesh",
      "type": "boilerplate/App/LOAD_POSTS_SUCCESS",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;
